src/framework/mlt_events.c

```diff
@@ -59,11 +59,20 @@ typedef struct mlt_events_struct *mlt_events;
 
 struct mlt_event_struct
 {
-	mlt_events owner;
+	mlt_events parent;
 	int ref_count;
 	int block_count;
 	mlt_listener listener;
-	void *service;
+	void *listener_data;
+};
+
+struct mlt_event_data_s {
```

Comment:
One suggestion to avoid the heap allocation and free() function: We could obfuscate this type and put it in the header. Then, perform the reinterpretation in the get/set functions by casting. For example:
```
struct mlt_event_data {
    union {
        int i;
        void* p;
    } u;
};
```
The size will be either 32 or 64 bits depending on the platform and could be passed by value instead of as a pointer.

The resulting functions would become:

```
mlt_event_data_set_int(mlt_event_data event_data, int value)
{
    event_data.i = int;
}

int mlt_event_data_get_int(mlt_event_data event_data)
{
    return event_data.i;
}

mlt_event_data_set_string(mlt_event_data event_data, const char *value)
{
    event_data.p = (void*)value;
}

const char *mlt_event_data_get_string(mlt_event_data event_data)
{
    return (const char*)event_data.p;
}
```


If we like this, maybe it should be a generic type defined in mlt_types named "mlt_opaque_data".

Reply:
I agree with the concerns about the heap allocations and will factor in your ideas except `mlt_opaque_data`.